// kepler.gl/dist/reducers/map-style.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapStyleReducerFactory = exports.INITIAL_MAP_STYLE = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _actionHandler; // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

// Actions


var _reduxActions = require('redux-actions');

var _actionTypes = require('../constants/action-types');

var _actionTypes2 = _interopRequireDefault(_actionTypes);

var _mapStyleUpdaters = require('./map-style-updaters');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// bedrock browserInit flattens our immutable object into a plain object
// we have to recreate the state after the app is loaded
var getDefaultState = function getDefaultState() {
  var visibleLayerGroups = {};
  var styleType = 'dark';
  var topLayerGroups = {};

  return {
    styleType: styleType,
    visibleLayerGroups: visibleLayerGroups,
    topLayerGroups: topLayerGroups,
    mapStyles: {},
    // save mapbox access token
    mapboxApiAccessToken: null,
    inputStyle: (0, _mapStyleUpdaters.getInitialInputStyle)()
  };
};

var INITIAL_MAP_STYLE = exports.INITIAL_MAP_STYLE = getDefaultState();

var actionHandler = (_actionHandler = {}, (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.INIT, _mapStyleUpdaters.initMapStyleUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.INPUT_MAP_STYLE, _mapStyleUpdaters.inputMapStyleUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.MAP_CONFIG_CHANGE, _mapStyleUpdaters.mapConfigChangeUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.MAP_STYLE_CHANGE, _mapStyleUpdaters.mapStyleChangeUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.LOAD_MAP_STYLES, _mapStyleUpdaters.loadMapStylesUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.LOAD_MAP_STYLE_ERR, _mapStyleUpdaters.loadMapStyleErrUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.RECEIVE_MAP_CONFIG, _mapStyleUpdaters.receiveMapConfigUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.LOAD_CUSTOM_MAP_STYLE, _mapStyleUpdaters.loadCustomMapStyleUpdater), (0, _defineProperty3.default)(_actionHandler, _actionTypes2.default.ADD_CUSTOM_MAP_STYLE, _mapStyleUpdaters.addCustomMapStyleUpdater), _actionHandler);

var mapStyleReducerFactory = exports.mapStyleReducerFactory = function mapStyleReducerFactory() {
  var initialState = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _reduxActions.handleActions)(actionHandler, (0, _extends3.default)({}, INITIAL_MAP_STYLE, initialState, { initialState: initialState }));
};

exports.default = mapStyleReducerFactory();
