// kepler.gl/dist/components/side-panel/panel-header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SaveExportDropdown = exports.PanelAction = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  background-color: ', ';\n  padding: 12px 16px 0 16px;\n'], ['\n  background-color: ', ';\n  padding: 12px 16px 0 16px;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  justify-content: space-between;\n  margin-bottom: 16px;\n  width: 100%;\n'], ['\n  display: flex;\n  justify-content: space-between;\n  margin-bottom: 16px;\n  width: 100%;\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n'], ['\n  display: flex;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  align-items: center;\n  border-radius: 2px;\n  color: ', ';\n  display: flex;\n  height: 26px;\n  justify-content: center;\n  margin-left: 4px;\n  width: 26px;\n\n  a {\n    height: 20px;\n  }\n\n  :hover {\n    cursor: pointer;\n    background-color: ', ';\n    color: ', ';\n\n    a {\n      color: ', ';\n    }\n  }\n'], ['\n  align-items: center;\n  border-radius: 2px;\n  color: ', ';\n  display: flex;\n  height: 26px;\n  justify-content: center;\n  margin-left: 4px;\n  width: 26px;\n\n  a {\n    height: 20px;\n  }\n\n  :hover {\n    cursor: pointer;\n    background-color: ', ';\n    color: ', ';\n\n    a {\n      color: ', ';\n    }\n  }\n']),
    _templateObject5 = (0, _taggedTemplateLiteral3.default)(['\n  background-color: ', ';\n  box-shadow: ', ';\n  font-size: 11px;\n  padding: 16px 0;\n  position: absolute;\n  transition: ', ';\n  display: flex;\n  margin-top: ', ';\n  opacity: ', ';\n  transform: translateX(calc(-50% + 20px));\n  pointer-events:  ', ';\n  z-index: 1000;\n\n  .save-export-dropdown__inner {\n    box-shadow: none;\n    background-color: transparent;\n    display: flex;\n  }\n\n  .save-export-dropdown__item {\n    align-items: center;\n    border-right: 1px solid ', ';\n    color: ', ';\n    display: flex;\n    flex-direction: column;\n    padding: 0 22px;\n\n    :hover {\n      cursor: pointer;\n      color: ', ';\n    }\n\n    &:last-child {\n      border-right: 0;\n    }\n  }\n\n  .save-export-dropdown__title {\n    white-space: nowrap;\n    margin-top: 4px;\n  }\n'], ['\n  background-color: ', ';\n  box-shadow: ', ';\n  font-size: 11px;\n  padding: 16px 0;\n  position: absolute;\n  transition: ', ';\n  display: flex;\n  margin-top: ', ';\n  opacity: ', ';\n  transform: translateX(calc(-50% + 20px));\n  pointer-events:  ', ';\n  z-index: 1000;\n\n  .save-export-dropdown__inner {\n    box-shadow: none;\n    background-color: transparent;\n    display: flex;\n  }\n\n  .save-export-dropdown__item {\n    align-items: center;\n    border-right: 1px solid ', ';\n    color: ', ';\n    display: flex;\n    flex-direction: column;\n    padding: 0 22px;\n\n    :hover {\n      cursor: pointer;\n      color: ', ';\n    }\n\n    &:last-child {\n      border-right: 0;\n    }\n  }\n\n  .save-export-dropdown__title {\n    white-space: nowrap;\n    margin-top: 4px;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents3 = require('../common/styled-components');

var _logo = require('../common/logo');

var _logo2 = _interopRequireDefault(_logo);

var _icons = require('../common/icons');

var _panelDropdown = require('./panel-dropdown');

var _panelDropdown2 = _interopRequireDefault(_panelDropdown);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StyledPanelHeader = _styledComponents2.default.div.attrs({
  className: 'side-side-panel__header'
})(_templateObject, function (props) {
  return props.theme.sidePanelHeaderBg;
});

var StyledPanelHeaderTop = _styledComponents2.default.div.attrs({
  className: 'side-panel__header__top'
})(_templateObject2);

var StyledPanelTopActions = _styledComponents2.default.div.attrs({
  className: 'side-panel__header__actions'
})(_templateObject3);

var StyledPanelAction = _styledComponents2.default.div.attrs({
  className: 'side-panel__header__actions'
})(_templateObject4, function (props) {
  return props.active ? props.theme.textColorHl : props.theme.subtextColor;
}, function (props) {
  return props.theme.secondaryBtnActBgd;
}, function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.textColorHl;
});

var StyledPanelDropdown = _styledComponents2.default.div(_templateObject5, function (props) {
  return props.theme.dropdownListBgd;
}, function (props) {
  return props.theme.dropdownListShadow;
}, function (props) {
  return props.theme.transitionSlow;
}, function (props) {
  return props.show ? '6px' : '20px';
}, function (props) {
  return props.show ? 1 : 0;
}, function (props) {
  return props.show ? 'all' : 'none';
}, function (props) {
  return props.theme.panelHeaderIcon;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.textColorHl;
});

var PanelAction = exports.PanelAction = function PanelAction(_ref) {
  var item = _ref.item,
      onClick = _ref.onClick;
  return _react2.default.createElement(
    StyledPanelAction,
    { className: 'side-panel__panel-header__action',
      'data-tip': true, 'data-for': item.id + '-action', onClick: onClick },
    _react2.default.createElement(
      'a',
      { target: item.blank ? '_blank' : '', href: item.href },
      _react2.default.createElement(item.iconComponent, { height: '20px' })
    ),
    _react2.default.createElement(
      _styledComponents3.Tooltip,
      {
        id: item.id + '-action',
        place: 'bottom',
        delayShow: 500,
        effect: 'solid'
      },
      _react2.default.createElement(
        'span',
        null,
        item.tooltip
      )
    )
  );
};

var PanelItem = function PanelItem(_ref2) {
  var onClose = _ref2.onClose,
      onClickHandler = _ref2.onClickHandler,
      label = _ref2.label,
      icon = _ref2.icon;
  return _react2.default.createElement(
    'div',
    { className: 'save-export-dropdown__item', onClick: function onClick(e) {
        e.stopPropagation();
        onClose();
        onClickHandler();
      } },
    icon,
    _react2.default.createElement(
      'div',
      { className: 'save-export-dropdown__title' },
      label
    )
  );
};

var SaveExportDropdown = exports.SaveExportDropdown = function SaveExportDropdown(_ref3) {
  var onExportImage = _ref3.onExportImage,
      onExportData = _ref3.onExportData,
      onExportConfig = _ref3.onExportConfig,
      onSaveMap = _ref3.onSaveMap,
      show = _ref3.show,
      onClose = _ref3.onClose;

  return _react2.default.createElement(
    StyledPanelDropdown,
    { show: show, className: 'save-export-dropdown' },
    _react2.default.createElement(
      _panelDropdown2.default,
      { onClose: onClose, className: 'save-export-dropdown__inner' },
      _react2.default.createElement(PanelItem, {
        label: 'Export Image',
        onClickHandler: onExportImage,
        onClose: onClose,
        icon: _react2.default.createElement(_icons.Picture, { height: '16px' })
      }),
      _react2.default.createElement(PanelItem, {
        label: 'Export Data',
        onClickHandler: onExportData,
        onClose: onClose,
        icon: _react2.default.createElement(_icons.Files, { height: '16px' })
      }),
      _react2.default.createElement(PanelItem, {
        label: 'Export Config',
        onClickHandler: onExportConfig,
        onClose: onClose,
        icon: _react2.default.createElement(_icons.CodeAlt, { height: '16px' })
      }),
      onSaveMap ? _react2.default.createElement(PanelItem, {
        label: 'Save Map Url',
        onClickHandler: onSaveMap,
        onClose: onClose,
        icon: _react2.default.createElement(_icons.Share, { height: '16px' })
      }) : null
    )
  );
};

var defaultActionItems = [{
  id: 'save',
  iconComponent: _icons.Save,
  tooltip: 'Save / Export',
  onClick: function onClick() {},
  dropdownComponent: SaveExportDropdown
}];

function PanelHeaderFactory() {
  var _class, _temp2;

  return _temp2 = _class = function (_Component) {
    (0, _inherits3.default)(PanelHeader, _Component);

    function PanelHeader() {
      var _ref4;

      var _temp, _this, _ret;

      (0, _classCallCheck3.default)(this, PanelHeader);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref4 = PanelHeader.__proto__ || Object.getPrototypeOf(PanelHeader)).call.apply(_ref4, [this].concat(args))), _this), _this.state = {
        dropdown: null
      }, _this.showDropdown = function (id) {
        _this.setState({ dropdown: id });
      }, _this.hideDropdown = function () {
        _this.setState({ dropdown: null });
      }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
    }

    (0, _createClass3.default)(PanelHeader, [{
      key: 'render',
      value: function render() {
        var _this2 = this;

        var _props = this.props,
            appName = _props.appName,
            version = _props.version,
            actionItems = _props.actionItems,
            onSaveMap = _props.onSaveMap,
            onExportImage = _props.onExportImage,
            onExportData = _props.onExportData,
            onExportConfig = _props.onExportConfig;

        return _react2.default.createElement(
          StyledPanelHeader,
          { className: 'side-panel__panel-header' },
          _react2.default.createElement(
            StyledPanelHeaderTop,
            { className: 'side-panel__panel-header__top' },
            _react2.default.createElement(this.props.logoComponent, { appName: appName, version: version }),
            _react2.default.createElement(
              StyledPanelTopActions,
              null,
              actionItems.map(function (item) {
                return _react2.default.createElement(
                  'div',
                  { className: 'side-panel__panel-header__right',
                    key: item.id, style: { position: 'relative' } },
                  _react2.default.createElement(PanelAction, {
                    item: item,
                    onClick: function onClick() {
                      if (item.dropdownComponent) {
                        _this2.showDropdown(item.id);
                      }

                      item.onClick();
                    }
                  }),
                  item.dropdownComponent ? _react2.default.createElement(item.dropdownComponent, {
                    onClose: _this2.hideDropdown,
                    show: _this2.state.dropdown === item.id,
                    onSaveMap: onSaveMap,
                    onExportData: onExportData,
                    onExportImage: onExportImage,
                    onExportConfig: onExportConfig
                  }) : null
                );
              })
            )
          )
        );
      }
    }]);
    return PanelHeader;
  }(_react.Component), _class.propTypes = {
    logoComponent: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.func]),
    actionItems: _propTypes2.default.arrayOf(_propTypes2.default.any)
  }, _class.defaultProps = {
    logoComponent: _logo2.default,
    actionItems: defaultActionItems
  }, _temp2;
}

exports.default = PanelHeaderFactory;
