// kepler.gl/dist/actions/ui-state-actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setExportFiltered = exports.setExportDataType = exports.setExportSelectedDataset = exports.cleanupExportImage = exports.setExportImageDataUri = exports.startExportingImage = exports.toggleLegend = exports.setResolution = exports.setRatio = exports.setExportData = exports.setExportConfig = exports.openDeleteModal = exports.toggleMapControl = exports.toggleModal = exports.toggleSidePanel = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _reduxActions = require('redux-actions');

var _actionTypes = require('../constants/action-types');

var _actionTypes2 = _interopRequireDefault(_actionTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var CLEANUP_EXPORT_IMAGE = _actionTypes2.default.CLEANUP_EXPORT_IMAGE,
    OPEN_DELETE_MODAL = _actionTypes2.default.OPEN_DELETE_MODAL,
    SET_EXPORT_DATA_TYPE = _actionTypes2.default.SET_EXPORT_DATA_TYPE,
    SET_EXPORT_FILTERED = _actionTypes2.default.SET_EXPORT_FILTERED,
    SET_EXPORT_IMAGE_DATA_URI = _actionTypes2.default.SET_EXPORT_IMAGE_DATA_URI,
    SET_EXPORT_CONFIG = _actionTypes2.default.SET_EXPORT_CONFIG,
    SET_EXPORT_DATA = _actionTypes2.default.SET_EXPORT_DATA,
    SET_EXPORT_SELECTED_DATASET = _actionTypes2.default.SET_EXPORT_SELECTED_DATASET,
    SET_RATIO = _actionTypes2.default.SET_RATIO,
    SET_RESOLUTION = _actionTypes2.default.SET_RESOLUTION,
    START_EXPORTING_IMAGE = _actionTypes2.default.START_EXPORTING_IMAGE,
    TOGGLE_LEGEND = _actionTypes2.default.TOGGLE_LEGEND,
    TOGGLE_MODAL = _actionTypes2.default.TOGGLE_MODAL,
    TOGGLE_SIDE_PANEL = _actionTypes2.default.TOGGLE_SIDE_PANEL,
    TOGGLE_MAP_CONTROL = _actionTypes2.default.TOGGLE_MAP_CONTROL;

// second argument of createAction is expected to be payloadCreator or undefined

var _map = [TOGGLE_SIDE_PANEL, TOGGLE_MODAL, TOGGLE_MAP_CONTROL, OPEN_DELETE_MODAL, SET_RATIO, SET_RESOLUTION, TOGGLE_LEGEND, START_EXPORTING_IMAGE, SET_EXPORT_IMAGE_DATA_URI, CLEANUP_EXPORT_IMAGE, SET_EXPORT_SELECTED_DATASET, SET_EXPORT_DATA_TYPE, SET_EXPORT_FILTERED, SET_EXPORT_CONFIG, SET_EXPORT_DATA].map(function (a) {
  return (0, _reduxActions.createAction)(a);
}),
    _map2 = (0, _slicedToArray3.default)(_map, 15),
    toggleSidePanel = _map2[0],
    toggleModal = _map2[1],
    toggleMapControl = _map2[2],
    openDeleteModal = _map2[3],

// export image
setRatio = _map2[4],
    setResolution = _map2[5],
    toggleLegend = _map2[6],
    startExportingImage = _map2[7],
    setExportImageDataUri = _map2[8],
    cleanupExportImage = _map2[9],

// export data
setExportSelectedDataset = _map2[10],
    setExportDataType = _map2[11],
    setExportFiltered = _map2[12],
    setExportConfig = _map2[13],
    setExportData = _map2[14];

exports.toggleSidePanel = toggleSidePanel;
exports.toggleModal = toggleModal;
exports.toggleMapControl = toggleMapControl;
exports.openDeleteModal = openDeleteModal;
exports.setExportConfig = setExportConfig;
exports.setExportData = setExportData;
exports.setRatio = setRatio;
exports.setResolution = setResolution;
exports.toggleLegend = toggleLegend;
exports.startExportingImage = startExportingImage;
exports.setExportImageDataUri = setExportImageDataUri;
exports.cleanupExportImage = cleanupExportImage;
exports.setExportSelectedDataset = setExportSelectedDataset;
exports.setExportDataType = setExportDataType;
exports.setExportFiltered = setExportFiltered;
